### PR TITLE
Fixes #282 Narrator does not convey error information when no more Right Parenthesis can be added in expression.

### LINKED ISSUE #282: Narrator does not convey error information when no more Right Parenthesis can be added in expression.
Test Environment:
OS Version: Windows 10 Build 17133.1
App Name: Calculator
App Version: 10.1804.911.0

Repro Steps: 

Launch "Calculator" app.
Launch Narrator.
Navigate to “Programmer/Scientific" control in the navigation pane and activate it.
Enter a number in the display.
Navigate to "Left Parenthesis" button in “Programmer/Scientific Calculator" window and activate it five times.
Now, navigate to "Right Parenthesis" button and activate it five times.
Try to activate "Right Parenthesis" button for the sixth time. No more Right Parenthesis can be added in expression now since the count of Left Parenthesis and Right Parenthesis is equal.
Verify if Narrator conveys error information when no more Right Parenthesis can be added in expression.
Actual Result:

Narrator does not convey error information when no more Right Parenthesis can be added in expression once the count of Left Parenthesis and Right Parenthesis is equal.       

Expected Result:

Narrator should convey error information when no more Right Parenthesis can be added in expression.

### PR DESCRIPTION
**Pull request by sanderl**
### Description of the changes:
- Removes left parenthesis only narrator announcment.
- Updates open parentheses count narrator announcement to be triggered when both right and left parenthesis buttons are clicked.
- Adds a narrator announcement for the case when there are no open parentheses and the right parenthesis button is clicked.

### How changes were validated:

- Manually tested the open parenthesis count narrator announcement occurred for left and right parenthesis clicks in Scientific and Programmer calculator modes.
- Manually tested the no open parenthesis narrator announcement occurred when the right parenthesis button was clicked and there were no open parentheses in Scientific and Programmer calculator modes.
- Tested the announcements were not triggered by clear in Scientific and Programmer calculator modes.
- Ran automated unit tests.

### CODE REVIEW

File: src/CalcManager/CalculatorManager.cpp [hunk 1]
```diff
@@ -117,6 +117,15 @@ namespace CalculationManager
         m_displayCallback->SetParenDisplayText(parenthesisCount);
     }
 
+    /// <summary>
+    /// Callback from the engine
+    /// Used to set the narrator text when no parenthesis can be added
```
Comment (danbelcher-MSFT) [inline src/CalcManager/CalculatorManager.cpp]: Let's not say what this is used for.  At this level of the code, we are simply bubbling an event and have no idea what the resulting behavior will be.

File: src/CalcManager/CalculatorManager.cpp [hunk 2]
```diff
@@ -117,6 +117,15 @@ namespace CalculationManager
...
+    /// </summary>
+    void CalculatorManager::OnNoParenAdded()
```
Comment (danbelcher-MSFT) [inline src/CalcManager/CalculatorManager.cpp]: Should this be `OnNoParenAdded` or `OnNoRightParenAdded`?  Are there cases where we don't add a left paren?

File: src/CalcViewModel/Common/Automation/NarratorAnnouncement.cpp [hunk 3]
```diff
@@ -142,3 +144,21 @@ NarratorAnnouncement^ CalculatorAnnouncement::GetDisplayCopiedAnnouncement(Strin
         AutomationNotificationKind::ActionCompleted,
         AutomationNotificationProcessing::ImportantMostRecent);
 }
+
+NarratorAnnouncement^ CalculatorAnnouncement::GetOpenParenethisCount(String^ announcement)
```
Comment (danbelcher-MSFT) [inline src/CalcViewModel/Common/Automation/NarratorAnnouncement.cpp]: `GetOpenParenthesisCountAnnouncement`?  `GetOpenParenthesisCountChangedAnnouncement`?

Comment (danbelcher-MSFT) [reply src/CalcViewModel/Common/Automation/NarratorAnnouncement.cpp]: Typo in the spelling for the current function name, `Parenethis`

File: src/CalcViewModel/Common/CalculatorDisplay.cpp [hunk 4]
```diff
@@ -49,6 +49,18 @@ void CalculatorDisplay::SetParenDisplayText(_In_ const std::wstring& parenthesis
     }
 }
 
+void CalculatorDisplay::OnNoParenAdded()
+{
+    if (m_callbackReference != nullptr)
+    {
+        auto calcVM = m_callbackReference.Resolve<ViewModel::StandardCalculatorViewModel>();
+        if (calcVM)
```
Comment (danbelcher-MSFT) [inline src/CalcViewModel/Common/CalculatorDisplay.cpp]: Even better is:
```C++
if (auto calcVM = ...)
{

}
```
This style has the exact same effect, except the `calcVM` variable is scoped to the body of the `if` block and does not exist outside, which is the best reflection of intent here.

File: src/CalcViewModel/Common/CalculatorDisplay.cpp [hunk 5]
```diff
@@ -49,6 +49,18 @@ void CalculatorDisplay::SetParenDisplayText(_In_ const std::wstring& parenthesis
...
+        {
+            calcVM->SetNoParenAddedNarratorAnnouncement();
```
Comment (danbelcher-MSFT) [inline src/CalcViewModel/Common/CalculatorDisplay.cpp]: Consistent with the other callbacks in this file, it seems more appropriate to call `calcVM->OnNoParenAdded()` and let the ViewModel internally decide that it should update the narrator announcement.

File: src/CalcViewModel/StandardCalculatorViewModel.cpp [hunk 6]
```diff
@@ -92,7 +93,9 @@ StandardCalculatorViewModel::StandardCalculatorViewModel() :
     m_localizedMemorySavedAutomationFormat(nullptr),
     m_localizedMemoryItemChangedAutomationFormat(nullptr),
     m_localizedMemoryItemClearedAutomationFormat(nullptr),
-    m_localizedMemoryCleared(nullptr)
+    m_localizedMemoryCleared(nullptr),
+    m_localizedOpenParenthesisCountChangedAutomationFormat(nullptr),
+    m_localizaedNoParenthesisAddedAutomationFormat(nullptr)
```
Comment (danbelcher-MSFT) [inline src/CalcViewModel/StandardCalculatorViewModel.cpp]: Typo. `localizaed`

File: src/CalcViewModel/StandardCalculatorViewModel.cpp [hunk 7]
```diff
@@ -222,10 +224,32 @@ void StandardCalculatorViewModel::SetParenthesisCount(_In_ const wstring& parent
     if (IsProgrammer || IsScientific)
     {
         OpenParenthesisCount = ref new String(parenthesisCount.c_str());
-        RaisePropertyChanged("LeftParenthesisAutomationName");
     }
 }
 
+void StandardCalculatorViewModel::OpenParenthesisCountNarratorAnnouncment()
```
Comment (danbelcher-MSFT) [inline src/CalcViewModel/StandardCalculatorViewModel.cpp]: The phrasing of this function name is somewhat ambiguous.  Probably, this should be `Set...` so it's clear that this is setting an announcement.

Comment (danbelcher-MSFT) [inline src/CalcViewModel/StandardCalculatorViewModel.cpp]: Typo. `Announcment`

File: src/CalcViewModel/StandardCalculatorViewModel.cpp [hunk 8]
```diff
@@ -222,10 +224,32 @@ void StandardCalculatorViewModel::SetParenthesisCount(_In_ const wstring& parent
...
+{
+    String^ parenthesisCount = ((m_OpenParenthesisCount == nullptr) ? "0" : m_OpenParenthesisCount);
+    wstring localizedParenthesisCount = std::wstring(parenthesisCount->Data());
```
Comment (danbelcher-MSFT) [inline src/CalcViewModel/StandardCalculatorViewModel.cpp]: Don't need explicit `std` namespace in cpps.  Can use `wstring localizedParenthesisCount = parenthesisCount->Data()`

File: src/CalcViewModel/StandardCalculatorViewModel.h [hunk 9]
```diff
@@ -284,12 +276,15 @@ namespace CalculatorApp
             void OnMemoryClear(_In_ Platform::Object^ memoryItemPosition);
             void OnPinUnpinCommand(Platform::Object^ parameter);
 
+            void OpenParenthesisCountNarratorAnnouncment();
```
Comment (danbelcher-MSFT) [inline src/CalcViewModel/StandardCalculatorViewModel.h]: Does this need to be `internal`, or can it be `private`?

File: src/Calculator/Resources/en-US/Resources.resw [hunk 10]
```diff
@@ -1313,14 +1313,18 @@
...
-    <value>Left parenthesis, open parenthesis count %1</value>
-    <comment>{Locked="%1"} Screen reader prompt for the Calculator "(" button on the scientific operator keypad. %1 is the localized count of open parenthesis, e.g. "2".</comment>
-  </data>
   <data name="closeParenthesisButton.[using:Windows.UI.Xaml.Automation]AutomationProperties.Name" xml:space="preserve">
     <value>Right parenthesis</value>
     <comment>Screen reader prompt for the Calculator ")" button on the scientific operator keypad</comment>
   </data>
+  <data name="Format_OpenParenthesisCountAutomationNamePrefix" xml:space="preserve">
+    <value>Open parenthesis count %1</value>
+    <comment>{Locked="%1"} Screen reader prompt for the Calculator "(" button on the scientific operator keypad. %1 is the localized count of open parenthesis, e.g. "2".</comment>
```
Comment (danbelcher-MSFT) [inline src/Calculator/Resources/en-US/Resources.resw]: Scientific and Programmer

File: src/CalcViewModel/Common/CalculatorDisplay.cpp (hunk 4 above)
Comment (sanderl) [reply src/CalcViewModel/Common/CalculatorDisplay.cpp]: I fixed this throughout the file.

---
In reply to: [264936066](https://github.com/Microsoft/calculator/pull/284#discussion_r264936066) [](ancestors = 264936066)

Comment (danbelcher-MSFT) [reply src/CalcViewModel/Common/CalculatorDisplay.cpp]: Thanks!

File: src/Calculator/Views/CalculatorProgrammerRadixOperators.xaml.cpp [hunk 11]
```diff
@@ -94,3 +98,11 @@ void CalculatorProgrammerRadixOperators::IsErrorVisualState::set(bool value)
         NumberPad->IsErrorVisualState = m_isErrorVisualState;
     }
 }
+
+void CalculatorProgrammerRadixOperators::OnPropertyChanged(Object^ sender, PropertyChangedEventArgs^ e)
+{
+    if (e->PropertyName == CalculatorViewModelProperties::OpenParenthesisCount && closeParenthesisButton->FocusState != Windows::UI::Xaml::FocusState::Unfocused)
```
Comment (danbelcher-MSFT) [inline src/Calculator/Views/CalculatorProgrammerRadixOperators.xaml.cpp]: You want to use `::FocusState::Unfocused` here.  The problem with `FocusState::Unfocused` is that `CalculatorProgrammerRadixOperators` is a `UserControl` so it has an inherited `FocusState` property. The compiler sees `FocusState::Unfocused` and thinks you are accessing that property, followed by `::Unfocused` which it doesn't understand.  The scope resolution operator, `::`, tells the compiler you are accessing the `FocusState` enumeration which exists in the global namespace after the `using Windows::UI::Xaml;` statement.
https://docs.microsoft.com/en-us/cpp/cpp/scope-resolution-operator


Same feedback for the other file.

File: src/Calculator/Views/CalculatorProgrammerRadixOperators.xaml.h [hunk 12]
```diff
@@ -33,8 +33,11 @@ namespace CalculatorApp
         void Shift_Clicked(Platform::Object^ sender, Windows::UI::Xaml::RoutedEventArgs^ e);
         void SetVisibilityBinding(Windows::UI::Xaml::FrameworkElement^ element, Platform::String^ path, Windows::UI::Xaml::Data::IValueConverter^ converter);
         void OnLoaded(Platform::Object^ sender, Windows::UI::Xaml::RoutedEventArgs^ e);
+        void OnUnloaded(Platform::Object^ sender, Windows::UI::Xaml::RoutedEventArgs^ e);
         void ProgModeRadixChange();
+        void OnPropertyChanged(Platform::Object^ sender, Windows::UI::Xaml::Data::PropertyChangedEventArgs ^ e);
```
Comment (danbelcher-MSFT) [inline src/Calculator/Views/CalculatorProgrammerRadixOperators.xaml.h]: Let's call this `OnViewModelPropertyChanged`. Same feedback for the other file.

File: src/CalcViewModel/StandardCalculatorViewModel.cpp [hunk 13]
```diff
@@ -226,6 +231,34 @@ void StandardCalculatorViewModel::SetParenthesisCount(_In_ const wstring& parent
...
+void StandardCalculatorViewModel::OnNoRightParenAdded()
+{
+    SetNoParenAddedNarratorAnnouncement();
+}
+
+void StandardCalculatorViewModel::SetNoParenAddedNarratorAnnouncement()
+{
+    String^ announcement = LocalizationStringUtil::GetLocalizedNarratorAnnouncement(
+        CalculatorResourceKeys::NoParenthesisAdded,
+        m_localizedNoRightParenthesisAddedAutomationFormat);
```
Comment (danbelcher-MSFT) [inline src/CalcViewModel/StandardCalculatorViewModel.cpp]: Should we drop `Automation` from this name?  I don't think this is used for an automation name.

File: src/Calculator/Resources/en-US/Resources.resw [hunk 14]
```diff
@@ -1313,14 +1313,22 @@
     <value>Left parenthesis</value>
     <comment>Screen reader prompt for the Calculator "(" button on the scientific operator keypad</comment>
   </data>
-  <data name="Format_OpenParenthesisAutomationNamePrefix" xml:space="preserve">
+    <data name="Format_OpenParenthesisAutomationNamePrefix" xml:space="preserve">
```
Comment (danbelcher-MSFT) [inline src/Calculator/Resources/en-US/Resources.resw]: re-align this with the `</data>` element above.

File: src/CalcManager/CalculatorManager.cpp (hunk 1 above)
Comment (sanderl) [reply src/CalcManager/CalculatorManager.cpp]: It looks like I removed it from the method above, but the comment applies to both methods, so now it is removed from both.

---
In reply to: [265407878](https://github.com/Microsoft/calculator/pull/284#discussion_r265407878) [](ancestors = 265407878)

**Pull request merged**